msal/src/main/java/com/microsoft/identity/nativeauth/statemachine/states/MFAStates.kt

```diff
@@ -0,0 +1,627 @@
+                        )
+                        MFARequestChallengeError(
+                            errorMessage = result.errorDescription,
+                            error = result.error,
+                            correlationId = result.correlationId,
+                            errorCodes = result.errorCodes,
+                            exception = result.exception
+                        )
+                    }
+                    is INativeAuthCommandResult.Redirect -> {
+                        MFARequestChallengeError(
+                            errorType = ErrorTypes.BROWSER_REQUIRED,
+                            error = result.error,
+                            errorMessage = result.errorDescription,
+                            correlationId = result.correlationId
+                        )
+                    }
+                }
+            } catch (e: Exception) {
+                MFARequestChallengeError(
+                    errorType = ErrorTypes.CLIENT_EXCEPTION,
+                    errorMessage = "MSAL client exception occurred in requestChallenge(authMethod).",
+                    exception = e,
+                    correlationId = correlationId
+                )
+            }
+        }
+    }
+
+    /**
+     * SubmitChallengeCallback receives the result for submitChallenge() in MFA flows in native authentication.
+     */
+    interface SubmitChallengeCallback : Callback<MFASubmitChallengeResult>
+
+    /**
+     * Submits the challenge value to the server; callback variant.
+     *
+     * <strong><u>Warning: this API is experimental. It may be changed in the future without notice. Do not use in production applications.</u></strong>
+     * @param callback [com.microsoft.identity.nativeauth.statemachine.states.MFARequiredState.SubmitChallengeCallback] to receive the result on.
+     * @return The result of the submit challenge action.
+     */
+    fun submitChallenge(challenge: String, callback: SubmitChallengeCallback) {
+        LogSession.logMethodCall(
+            tag = TAG,
+            correlationId = correlationId,
+            methodName = "${TAG}.submitChallenge(callback: SubmitChallengeCallback)"
+        )
+        NativeAuthPublicClientApplication.pcaScope.launch {
+            try {
+                val result = submitChallenge(challenge)
+                callback.onResult(result)
+            } catch (e: MsalException) {
+                Logger.error(TAG, "Exception thrown in submitChallenge", e)
+                callback.onError(e)
+            }
+        }
+    }
+
+    /**
+     * Submits the challenge value to the server; Kotlin coroutines variant.
+     *
+     * <strong><u>Warning: this API is experimental. It may be changed in the future without notice. Do not use in production applications.</u></strong>
+     * @return The result of the submit challenge action.
+     */
+    suspend fun submitChallenge(challenge: String): MFASubmitChallengeResult {
```

Comment:
on iOS we check that the challenge is not empty. If it is, we return "invalidChallenge" error

Reply:
We don't do that kind of validation anywhere (also not in submitCode in sign in, sign up and SSPR flows). We do this in the API layer (`SignInTokenRequest`), where we're re-using the existing signInSubmitCode flows.